src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -230,6 +231,65 @@ master_drop_all_shards(PG_FUNCTION_ARGS)
 
 
 /*
+ * master_drop_sequences attempts to drop a list of sequences on a specified
+ * node. The "IF EXISTS" clause is used to permit dropping sequences even if
+ * they may not exist. Returns true on success, false on failure.
+ */
+Datum
+master_drop_sequences(PG_FUNCTION_ARGS)
+{
+	ArrayType *sequenceNamesArray = PG_GETARG_ARRAYTYPE_P(0);
+	text *nodeText = PG_GETARG_TEXT_P(1);
+	int64 nodePort = PG_GETARG_INT64(2);
+	bool dropSuccessful = false;
+	char *nodeName = TextDatumGetCString(nodeText);
+
+	ArrayIterator sequenceIterator = NULL;
+	Datum sequenceText = 0;
+	bool isNull = false;
+
+	StringInfo dropSeqCommand = makeStringInfo();
+
+	/* this runs from a trigger that must run as superuser */
+	if (!superuser())
+	{
+		ereport(ERROR, (errmsg("cannot drop sequences of a dropped table as "
+							   "non-superuser")));
+	}
+
+	sequenceIterator = array_create_iterator(sequenceNamesArray, 0, NULL);
+	while (array_iterate(sequenceIterator, &sequenceText, &isNull))
+	{
+		if (isNull)
+		{
+			ereport(ERROR, (errmsg("unexpected NULL sequence name"),
+							errcode(ERRCODE_INVALID_PARAMETER_VALUE)));
+		}
+
+		if (dropSeqCommand->len == 0)
```

Comment:
On 2016-07-27 14:18:06 -0700, Jason Petersen wrote:
> > +	if (!superuser())
> > +	{
> > +		ereport(ERROR, (errmsg("cannot drop sequences of a dropped table as "
> > +							   "non-superuser")));
> > +	}
> > +
> > +	sequenceIterator = array_create_iterator(sequenceNamesArray, 0, NULL);
> > +	while (array_iterate(sequenceIterator, &sequenceText, &isNull))
> > +	{
> > +		if (isNull)
> > +		{
> > +			ereport(ERROR, (errmsg("unexpected NULL sequence name"),
> > +							errcode(ERRCODE_INVALID_PARAMETER_VALUE)));
> > +		}
> > +
> > +		if (dropSeqCommand->len == 0)
> 
> My own pathology: I _really_ can't stand the "are we on the first/last iteration?" variables, and I rewrote this to avoid having to add another variable in the enclosing scope. I'll add a comment?

But that's just that? Just in a harder to understand version.